hasktorch/src/Torch/Typed/NN/Recurrent/LSTM.hs

```diff
@@ -0,0 +1,478 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NoStarIsType          #-}
+{-# LANGUAGE OverloadedLists       #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE StandaloneDeriving    #-}
+{-# LANGUAGE StrictData            #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.Normalise #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.Extra.Solver #-}
+
+module Torch.Typed.NN.Recurrent.LSTM
+    (LSTMSpec(..)
+    , Directionality(..)
+    , NumberOfDirections
+    , LSTM(..)
+    , LSTMParams(..)
+    , ParamsPerDirection(..)
+    , forwardNoDropout
+    , forwardWithDropout)
+where
+
+import qualified ATen.Cast                as ATen
+import qualified ATen.Class               as ATen
+import qualified ATen.Managed.Type.Tensor as ATen
+import qualified ATen.Type                as ATen
+import           Data.Kind
+import           Foreign.ForeignPtr
+import           GHC.Generics
+import           GHC.TypeLits
+import           GHC.TypeLits.Extra
+import           Prelude                  hiding (tanh)
+import           System.Environment
+import           System.IO.Unsafe
+import qualified Torch.Autograd           as A
+import qualified Torch.Device             as D
+import qualified Torch.DType              as D
+import qualified Torch.Functions          as D
+import qualified Torch.NN                 as A
+import qualified Torch.Tensor             as D
+import qualified Torch.TensorFactories    as D
+import           Torch.Typed
+import           Torch.Typed.Factories
+import           Torch.Typed.Native       (expand, lstm)
+import           Torch.Typed.NN
+import           Torch.Typed.Tensor
+
+
+-- | A specification for a long, short-term memory layer.
+--
+data LSTMSpec device (dir :: Directionality)
+        (dtype :: D.DType)
```

Comment:
In my `nn` code, I decided to put `dtype` and `device` at the end, cf. https://github.com/hasktorch/hasktorch/blob/master/hasktorch/src/Torch/Typed/NN.hs#L89, for instance.
the reasoning was that, with type applications enabled, one would almost never need to apply these, but almost always have to annotate the others.

Reply:
Good call